File: app/Asteroids.io/Asteroids.io/Command/CommandBuffer.swift
//
//  CommandBuffer.swift
//  Asteroids.io
//
//  Created by Tony on 4/24/19.
//  Copyright © 2019 cmsc436. All rights reserved.
//

import Foundation

class CommandBuffer {
    var buffer: [Command]
    var length: Int
    
    
    init(length: Int) {
        self.length = length
        self.buffer = []
    }
    
    func add(command: Command) {
        if (buffer.count < length) {
            buffer.append(command)
        }
        else {
            buffer.removeFirst()
            buffer.append(command)
        }
    }
    
    func revertTo(frame: Int, f: (Command) -> Void) {
        let currentFrame = GameManager.default.frame
        let index = currentFrame - frame
        
        for i in stride(from: buffer.count - 1, to: index, by: -1) {
            f(buffer[i])
        }
    }
    
    func simulateFrom(frame: Int, f: (Command) -> Void) {
        let currentFrame = GameManager.default.frame
        let index = currentFrame - frame
        
        for i in index..<buffer.count {
            f(buffer[i])
        }
    }
}

File: server/Netcode/Command/command-buffer.js
module.exports = class CommandBuffer {

    constructor(length) {
        this.length = length;
        this.buffer = [];
        this.frame = 0;
    }

    setFrame(frame) {
        this.frame = frame
    }

    addCommand(command) {
        if (this.buffer.length < this.length) {
            this.buffer.push(command);
        }
        else {
            this.buffer.shift();
            this.buffer.push(command);
        }
        this.frame++;
    }

    insertCommand(command) {
        var index = (this.buffer.length - 1) - (this.frame - command.frame);
        if (index > 0) {
            this.buffer[index] = command;
        }
    }

    simulateFrom(frame, func) {
        var from = (this.buffer.length - 1) - (this.frame - frame);
        var to = this.buffer.length;

        for (from; from < to; from++) {
            func(this.buffer[from]);
        }
    }

    revertTo(frame, func) {
        console.log(this.frame + ' ' + frame);
        var to = (this.buffer.length - 1) - (this.frame - frame);
        var from = this.buffer.length - 1;
        for (from; from >= to; from--) {
            console.log(this.buffer[from]);
            func(this.buffer[from]);
        }
    }

    getCommandAt(frame) {
        return this.buffer[this.frame - frame];
    }
}
